File: 백준/Silver/9012. 괄호/괄호.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const n = Number(input[0]);
// 결과 : YES or NO
let result = [];

for(let i = 1; i<=n; i++) {
    const data = input[i];
    let stack = [];
    
    // 각 예제들에 대해서 문자열 순회
    for(const str of data) {
        if(str === '(') { 
            // 여는 괄호를 스택에 추가
            stack.push(str); 
        } else if(str === ')') {
            // 닫힌 괄호 : 만약 스택이 비어있으면 -> 괄호쌍이 안 만들어짐 -> NO
            if(stack.length === 0) {
                result.push("NO");
                break; // 반복문 탈출
            }
            stack.pop(); // 닫힌 괄호와 쌍을 이루는 여는 괄호 제거
        }
    }
    
    // 결과처리
    if(stack.length === 0 && result.length < i) {
        result.push("YES"); // 스택이 비어있고, 결과 배열에 추가가 안된 상태
    } else if(result.length < i) {
        result.push("NO"); // 스택이 안 비어있는데, 결과 배열에 추가가 안된 상태
    }
}


console.log(result.join('\n'));

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

N = int(input().strip())

for _ in range(N):
    is_all_pair = True
    
    stack = []
    ps = list(input().strip())
    for ps_string in ps:
        if ps_string == '(':
            stack.append(ps_string)
        else:
            if stack:
                stack.pop()
            else:
                is_all_pair = False
                break

    # 여전히 스택에 남아있다면
    if stack:
        is_all_pair = False

    # 결과 출력
    print('YES' if is_all_pair == True else 'NO')